### ISSUE TITLE
Callback method and map access

### ISSUE THREAD

**Issue by kipusoep** [Joniff/Terratype#9]
Hi, great package!
Looking for a way to hook into the callback method, wouldn't be too hard to add I guess, right?
It would be great if the map object can be accessed too.

We're using the Google Maps v3 plugin.

**Comment by Joniff** [comment]
Hi Kip,

I have no plans for adding callback or events into Terratype when it renders in the backend, eg. When the maps are being viewed by content editors in Umbraco.

I have been asked a few times for allowing access to the maps when one is being rendered in a razor template on a front end page. I agree, it sounds simple and I might just add a horrible hack for those who are prepared to write js code that works only for GMaps. 

What I wanted to do though was to have a solution that would work for all Terratype providers, so you could write the same js code that worked for any or all of GMaps, Bing Maps and Leaflet maps. Or combinations of all three if you had some really weird front-end page with lots of different maps present. It is this, that I want to write, which is why I've been hesitant to start.

Cheers for the feedback and thank you.

tl;dr It is on my Roadmap. I do have plans to create a new release in the next few weeks, I might add a hack for those who can't wait for a complete/proper solution.

**Comment by kipusoep** [comment]
Nice! I meant for the front-end indeed.
Reason why I'd want this is to be able to show the infowindow on load, this might be even better as an extra option in the datatype settings :-)

**Comment by Joniff** [comment]
I really have to thank you, because this is the sort of thing that Terratype should be doing out-of-the-box without you having to resort to js and I hadn't thought of.

@Html.Terratype() has a Options argument that could have another property saying whether you want the label (info window) to be auto displayed. I think this could be done for all three Map providers relatively quickly, and would be something that isn't a hack.

This would be quick to implement and I agree, would be very useful.

Cheers for this, I will add this separately to the roadmap.

**Comment by geoffbeaumont** [comment]
I can understand wanting to create a consistent API for interacting with Terratype maps, but it may be overthinking this.

I'd say most of us are happy to select our map provider and write custom code addressing only that. I can't think of any occasion in 18 years of web development when I've needed to migrate map functionality from one provider to another (not saying it doesn't happen - but I've never personally had to).

If you provide access to the providers call back and native map object then anything that can be done with the APIs of that provider can be done with Terratype - if you try to wrap everything up in your own API then you have two significant problems:

1. You have to figure out how to handle differences in the feature sets available. If Bing offers feature A and Google doesn't, do you exclude that feature from your API? Or include it, but then what do you do if it's called and the map provider is Google?
2. You'll constantly be playing catch-up. How are you going to handle timely addition of new features/changes to the provider APIs?

If you provide access to the raw provider interfaces then you could still deliver a common Terratype API for certain features, but developers are not limited to this if they want to do something you haven't provided for.

Another angle - as a developer, I know my way round the Google Maps API, I use it on multiple frameworks and I often need to integrate code targetting Google Maps API written by other developers. I don't actually want to have to learn a different API for mapping on Umbraco, nor do I want to have to rewrite third party code to a Terratype API. Yes, I can use the Terratype data to render standard Google Maps code (which I'd have to at the moment), but Terratype provides a nice clean way of rendering the base map, and it's a shame we can't use that if we want to do anything remotely advanced.

**Comment by Joniff** [comment]
I enjoyed your impassioned words about wanting a, what is in my words, "a horrible hack for those who are prepared to write js code that works only for GMaps". 

To move the conversation along, as I suppose repeating that I do want Terratype to stay map provider agnostic, and I do consider it important.

Is how do I write this hack (Knowing that you can have multiple different maps being rendered at any one time).

1. Some global accessible variable, that is an array of maps currently being rendered by @Html.Terratype(). The downside is that this array will only have valid values after some undetermined period of time - eg, when all the maps have been rendered. 

2. Add a javascript function name value to Options, that is called when the map has finished rendering. That function's first argument is the gmap object for that map.

3. Implement some sort of Terratype js client service library, but for now have only one method that allows you to hook into some sort of promise that is invoked when the map has finished rendering, which returns the Gmap object. Maybe later have two other versions of this library for easy Angular and React integration.

And there you see my problem, each hack solution requires more code.

I'm happy to accept a PR.

**Comment by geoffbeaumont** [comment]
1. Sounds horrible. You'd end up having to do polling until a valid state was detected.
2. Could work - fundamentally it's what you do with native Google Maps anyway. Plus it would work for at least Bing Maps too, just pass the Bing map object instead (i.e. the contract would be simply to return a JavaScript object, which would be the native map object for the underlying provider). I guess it might be worth passing an identifier of the provider in use, too, although that could probably be determined by checking the object type.
3. Is at this point much the same as 2, but I guess is the more elegant method which you're far less likely to be itching to deprecate and get rid of later!

I'd actually quite like to have a look at this, but I'm not sure when I'm going to get the time - if I do get a chance I'll let you know first in case you've already started on it.

**Comment by JohnBergman** [comment]
I like the idea of some sort of eventhooks that could be used to customize or override behavior, rendering, etc.

**Comment by lars-erik** [comment]
Reg. 1.  
Make a private deferred object which is resolved when all maps are rendered.  
https://api.jquery.com/deferred.promise/  
Make a globally accessible function that returns the promise.

Then we can go:

    $.when(terratype.isDone(function(maps) {
        // maps is an object with stuff or just an array of the provider map types.
        // could always just reject the promise for providers that lack support
        // ie. throw NotImplementedException();
    })

**Comment by Joniff** [comment]
Liking the solution as its lookign quick to do, is it possible though, @lars-erik, to have something similar, but not dependant on jquery though ?
Can I use http://es6-features.org/#PromiseUsage to do something similar ?

**Comment by lars-erik** [comment]
Take your pick of polyfill. :)  
https://ourcodeworld.com/articles/read/316/top-5-best-javascript-promises-polyfills

**Comment by geoffbeaumont** [comment]
How safe is it going to be to deploy polyfills within a package designed for use on other peoples sites? Need to be careful that it successfully detects functionality that's already supplied by an existing polyfill and doesn't load. Might be a non issue, but layering multiple duplicate polyfills on the same site isn't something I've ever tried.

**Comment by lars-erik** [comment]
Polyfills are usually distributed with an existence check. Then it's just a matter of inclusion order, so it should be quite safe. Another option would be to copy one into the closure to keep it super compatible. (IE. roll-your-own)

**Comment by Joniff** [comment]
New for release 1.0.15

Have added a javascript event object. See manual for full details, but there is now a global terratype javascript object that can be used to hook into events and control maps.

**Comment by kipusoep** [comment]
Thanks @Joniff, works great!

For reference, this is the code I'm using to open the infoWindow on load:
```js
(function() {
    if (typeof window.terratype !== "undefined") {
        window.terratype.onRender(function (provider, map) {
            if (typeof map.positions !== "undefined" && map.positions.length) {
                var poi = map.positions[0];
                if (typeof poi !== "undefined" && poi != null && typeof window.google !== "undefined" && window.google != null && typeof window.google.maps !== "undefined" && window.google.maps != null) {
                    window.google.maps.event.trigger(poi.handle, "click");
                }
            }
        });
    }
})();
```

**Comment by Joniff** [comment]
@kipusoep While your code looks great, there is an easier way to preload the InfoWindow, in the options for the Html.Terratype() code, set AutoShowLabel = true.

`@Html.Terratype(new Options {AutoShowLabel = true}, map)`

**Comment by kipusoep** [comment]
That would be too easy ;-)
Why didn't you say so as reply on https://github.com/Joniff/Terratype/issues/9#issuecomment-335839842? :-p

**Comment by Joniff** [comment]
I really should have re-read the whole thread. I can only apologise.

**Comment by kipusoep** [comment]
No problem! Works great and it's easier to extend the map this way.

**Comment by lars-erik** [comment]
#H5YR! :)
